### PR TITLE
Add ability to override item's point-range using meta

### LINKED ISSUE #10164: Add per-ItemStack point range
## Problem

This is a problem. As a modder, I and probably many others as well, have wanted a way to change the range (distance for digging and placing nodes, and punching) of the hand tool at runtime. However, it can only be set by overriding the hand tool while loading a mod. I know this isnt something a whole lot of nodders may take advantage of, but it could definitely help with custom games.

## Solutions

For stacks and stack meta refs, you can use

ItemStack:get_toolcapabilities()
 and
ItemStackMetaRef:set_toolcapabilities(tool_caps)

IT would be nice to also have

ItemStack:get_range()
 and
ItemStack:set_range(range) --or nil to reset it to default

## Additional context

The point of this is also to be able to change the range of an ItemStack at runtime.
I know that you could just create a tool that has a farther range but this is very different and does not solve the problem. Please consider this. Thanks.

### PR DESCRIPTION
**Pull request by LandonAConway**
This PR adds the ability to override item point-range via ItemStack meta and fixes #10164 :

`meta:set_string("range", "8")`

## To do

This PR is Ready for Review.

## How to test

In devtest, use the Item Meta Editor to edit a tool, and add the "range" key then set it to a number.

## Notes

For some reason, the files contributed cannot be compared without looking at all of the code. I'm not really sure why this is happening but I'm trying to fix it. If anyone knows of a way to fix it, please let me know.

EDIT: The line endings have been fixed.

### CODE REVIEW

File: src/inventory.cpp [hunk 1]
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
+    std::string sRange = metadata.getString("range");
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
+    {
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
+			char *c = const_cast<char *>(sRange.c_str());
```
Comment (appgurueu) [inline src/inventory.cpp:285]: I don't see why `c` is not a `const char *`, and I don't see why the `const_cast` would be necessary:

* `c_str` has the signature [`const char* c_str() const;`](https://cplusplus.com/reference/string/string/c_str/)
* `strtof` has the signature [`float strtof( const char* str, char** str_end );`](https://en.cppreference.com/w/cpp/string/byte/strtof)

File: src/inventory.cpp [hunk 2]
```diff
@@ -271,6 +271,29 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
+    {
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'
+			f32 _range = std::strtof(sRange.c_str(), NULL);
+            if (_range != 0 || sRange == "0")
```
Comment (appgurueu) [inline src/inventory.cpp:286]: This looks like a weird way to do validation:

* You will reject something like `0e42` or `0.0`, despite it being a valid float literal.
* It is not clear that the `_range != 0` is actually checking for an error.
* All of this is inside a `try`, implying that this was doing all the error handling (which it doesn't)

Also, the overall logic confuses me: You return range -1 for an "invalid" range, but you return the itemdef range for an empty range. Shouldn't you just default to the itemdef range in both cases?

Side note: Please format this to comply with the [code style guidelines](https://dev.minetest.net/Code_style_guidelines).

File: src/inventory.h [hunk 3]
```diff
@@ -51,6 +51,9 @@ struct ItemStack
 	std::string getDescription(const IItemDefManager *itemdef) const;
 	std::string getShortDescription(const IItemDefManager *itemdef) const;
 
+    // Returns the range
```
Comment (appgurueu) [inline src/inventory.h:54]: This comment is pretty redundant.

File: src/inventory.cpp (hunk 2 above)
Comment (LandonAConway) [reply src/inventory.cpp:286]: > This looks like a weird way to do validation:
> 
> * You will reject something like `0e42` or `0.0`, despite it being a valid float literal.
> * It is not clear that the `_range != 0` is actually checking for an error.
> * All of this is inside a `try`, implying that this was doing all the error handling (which it doesn't)
> 
> Also, the overall logic confuses me: You return range -1 for an "invalid" range, but you return the itemdef range for an empty range. Shouldn't you just default to the itemdef range in both cases?
> 
> Side note: Please format this entire PR to comply with the [code style guidelines](https://dev.minetest.net/Code_style_guidelines).

I will remove the try-catch block if I can do it the way I'm about to explain. I want to create a function that returns false if the string is not a valid float before converting it. This was my original plan. The function would output like this.

`"Jdh76" -> false`
`"76.9" -> true`
`"836hdg" -> false`
`" " -> false`

Creating this kind of check function would be really simple. The problem is that I would have to place it in the inventory.cpp file which doean't seem like a very organized aproach but rather messy, so then where would I put it?

**Comment by LandonAConway** [comment]
I changed the code in `inventory.cpp` so `ItemStack::getRange()` will now have a helper function to check if the string is a valid number so it can be converted to a float and while doing so, it will also convert the string to a float.

I also changed the way `getToolRange()` works (inside `tool.h` and `tool.cpp`) and updated the code which uses that function to comply with the new arguments. I felt that this was neccecary because `getToolRange()` determines which range should be used between the player's hand and the selected item. I.e. if the player has nothing in their hand, then another range will be used. This range from the actual player hand should also use the modified `range` in the ItemStack's metadata. This makes sense and it would not make sense to do it another way. Since `getToolRange()` originally required two arguments of `ItemDefenition`, it could only get the range from those definitions. Instead it now requires two `f32` types representing the player's hand and selected item's ranges which should be retrieved from `ItemStack::getRange()`.

**Review by JosiahWI — CHANGES_REQUESTED** [review]

File: src/inventory.cpp [hunk 4]
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
+    {
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
+			char *c = const_cast<char *>(sRange.c_str());
+			f32 _range = std::strtof(c, NULL);
```
Comment (JosiahWI) [inline src/inventory.cpp:286]:
```suggestion
			f32 _range = std::strtof(sRange.c_str(), nullptr);
```
`const_cast` is a special feature and should be treated with caution and respect.

File: src/inventory.cpp [hunk 5]
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string sRange = metadata.getString("range");
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
+    {
+        try {
```
Comment (JosiahWI) [inline src/inventory.cpp:282]: `std::strtof` doesn't throw on parse failure; you don't need this. `std::stof` does throw, however.

File: src/inventory.cpp (hunk 4 above)
Comment (JosiahWI) [inline src/inventory.cpp:286]: I would recommend adding the `<cstdlib>` header for `std::strtof`.

File: src/inventory.cpp [hunk 6]
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
 	return desc;
 }
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string sRange = metadata.getString("range");
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
```
Comment (JosiahWI) [inline src/inventory.cpp:280]:
```suggestion
    else {
```

File: src/inventory.cpp [hunk 7]
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
+    {
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
```
Comment (JosiahWI) [inline src/inventory.cpp:284]: Did you try including the `<string>` header where `std::stof` is declared?

**Review by JosiahWI — CHANGES_REQUESTED** [review]

**Comment by LandonAConway** [comment]
Is there a way to remove certian commits from this pr? I updated the line endings settings in .gitattributes but I do not want to push that to official Minetest, and still, my files have the wrong line endings.

**Comment by JosiahWI** [comment]
`git rebase` is the tool you are looking for. I recommend using the interactive version (`git rebase -i`), especially if you haven't used the rebase command before.

**Comment by fluxionary** [comment]
i haven't played w/ this pull, but i see a potential confound. how does this interact w/ the capabilities of the player's hand, the `""` item?

File: src/client/game.cpp [hunk 8]
```diff
@@ -3245,7 +3245,7 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud)
 	const ItemStack &tool_item = player->getWieldedItem(&selected_item, &hand_item);
 
 	const ItemDefinition &selected_def = selected_item.getDefinition(itemdef_manager);
```
Comment (JosiahWI) [inline src/client/game.cpp:3247]:
```suggestion
```
Looks to me like you don't need this temporary anymore.

File: src/inventory.cpp [hunk 9]
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+    {
+        is_valid = false;
+        return 0;
+    }
+
+    char* ptr;
+    float f = strtof(str.c_str(), &ptr);
+    is_valid = (*ptr) == '\0';
+    return f;
+}
```
Comment (JosiahWI) [inline src/inventory.cpp:45]: It'd be a lot cleaner to get `std::stof` working.

File: src/inventory.cpp [hunk 10]
```diff
@@ -271,6 +286,17 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
 	return desc;
 }
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string str_meta_range = metadata.getString("range");
```
Comment (JosiahWI) [inline src/inventory.cpp:291]: This should use a reference, probably a const reference.

File: src/inventory.cpp [hunk 11]
```diff
@@ -271,6 +286,17 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
...
+    f32 range = getDefinition(itemdef).range;
+    bool is_valid;
+    float meta_range = string_to_float(str_meta_range, is_valid);
+    if (is_valid && meta_range > -1)
```
Comment (JosiahWI) [inline src/inventory.cpp:295]: Why is there no warning about the invalid range value?

**Review by JosiahWI — CHANGES_REQUESTED** [review]

**Comment by LandonAConway** [comment]
> `git rebase` is the tool you are looking for. I recommend using the interactive version (`git rebase -i`), especially if you haven't used the rebase command before.

I see exactly what you are talking about and it has been fixed already to handle this. (should be).
In `tool.cpp` there is a function that is called `getToolRange()`. It has been changed to take two floats (one for the selected item, and one for the hand) and returns the one to be used. This function once took two item definitions (one for the selected item, and hand) then got the range from the definitions but it has been changed to evaluate two floats. Both floats are retrieved from `ItemStack::getRange()` which gets the item definition's range, and if the metadata has a value at the key of `"range"` that is convertable to a float, it uses this instead.

File: src/client/game.cpp (hunk 8 above)
Comment (LandonAConway) [reply src/client/game.cpp:3247]: I actually did remove it. If you go down further into this function, the `selected_def` reference is actually used. Should it be moved down to where it is used?

Comment (JosiahWI) [reply src/client/game.cpp:3247]: Yes.

File: src/inventory.cpp (hunk 10 above)
Comment (LandonAConway) [reply src/inventory.cpp:291]: As I'm new to C++, I'm not exactly sure what this means. What value are you referring to that should be a const? It seems that they are all const. The function is a const and so is the IItemDefManager.

File: src/inventory.cpp (hunk 11 above)
Comment (LandonAConway) [reply src/inventory.cpp:295]: Are you asking to have a comment added for a warning?

File: src/inventory.cpp (hunk 10 above)
Comment (JosiahWI) [reply src/inventory.cpp:291]: I'm sorry, I was unclear. I'm referring to str_meta_range; copying strings is expensive.
```suggestion
    const std::string &str_meta_range = metadata.getString("range");
```

File: src/inventory.cpp (hunk 11 above)
Comment (JosiahWI) [reply src/inventory.cpp:295]: I'm asking why nothing is printed to the log. Invalid meta seems like something that should generate at least a warning.

**Comment by Desour** [comment]
The diffs are broken.

Also docs are missing.

**Comment by LandonAConway** [comment]
> The diffs are broken.
> 
> Also the documentation is missing.

I will get back to this PR hopefully in a couple days. I'm busy at work. I WILL be working on it so if there is inactivity for a couple days, just beware. This PR is really important to me.

**Comment by JosiahWI** [comment]
serverpackethandler.cpp has been severely afflicted by these changes.

**Comment by LandonAConway** [comment]
> serverpackethandler.cpp has been severely afflicted by these changes.

Literally only one line has changed. I have updated ALL the file's line endings so hopefully that will work.

File: src/inventory.cpp [hunk 12]
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/serialize.h"
 #include "util/string.h"
 
+//used in ItemStack::getRange()
+static float string_to_float(const std::string& str, bool& is_valid)
+{
+    if (str.empty())
+    {
+        is_valid = false;
+        return 0.0;
```
Comment (JosiahWI) [inline src/inventory.cpp:38]: `0.0f`, `0.0` is a `double` literal.

File: src/inventory.cpp [hunk 13]
```diff
@@ -271,6 +286,19 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
 	return desc;
 }
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string str_meta_range = metadata.getString("range");
```
Comment (JosiahWI) [inline src/inventory.cpp:291]: There is still an unnecessary string copy here.

File: src/inventory.cpp [hunk 14]
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/serialize.h"
 #include "util/string.h"
 
+//used in ItemStack::getRange()
+static float string_to_float(const std::string& str, bool& is_valid)
```
Comment (JosiahWI) [inline src/inventory.cpp:33]: We can use `std::optional` now that we have C++17 support, instead of having a boolean out parameter.

**Comment by LandonAConway** [comment]
I'm aware of the changes needed. I willbe working on them this week (hopefully).

**Pull request closed without merging**